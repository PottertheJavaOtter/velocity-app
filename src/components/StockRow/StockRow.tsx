import React from 'react'
import { IonItem, IonAvatar, IonLabel, IonNote } from '@ionic/react'
import { formatDecimalToDollars } from '../../util/moneyFormatter'

interface StockRowProps {
  /** The id of this stock entry */
  id: string;
  /** The image url of the stock */
  avatar?: string;
  /** Name of the company */
  name: string;
  /** Stock ticker symbol */
  ticker: string;
  /** Number of shares owned */
  shares: number;
  /** The annual dividend amount per share*/
  dividendPerShare: number;
}
const StockRow: React.FC<StockRowProps> = ({ id, name, ticker, shares, dividendPerShare = 0, avatar = "https://picsum.photos/100/100" }) => {
  const annualDividnd = formatDecimalToDollars(dividendPerShare * shares);
  return (
    <IonItem href={`/stock/${id}`} data-testid={`stock-row-${ticker}`}>
      <IonAvatar slot="start">
        <img src={avatar} />
      </IonAvatar>
      <IonLabel>
        <h2>{`${name} (${ticker})`}</h2>
        <p>Shares: {shares}</p>
      </IonLabel>
      <IonNote slot="end" color="primary">{annualDividnd}</IonNote>
    </IonItem>
  )
}

export default StockRow
