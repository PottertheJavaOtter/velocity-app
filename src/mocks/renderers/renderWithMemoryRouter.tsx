// @ts-nocheck
import React from "react";
import { render } from "@testing-library/react";
import { MemoryRouter, Route, Router } from "react-router-dom";
import { History, LocationDescriptor, createMemoryHistory } from "history";

// const renderWithMemoryRouter = (
//   ui: React.ReactNode,
//   initialEntries: LocationDescriptor[] = ["/my/initial-route"]
// ) => {
//   let historyInTest: History;
//   return {
//     ...render(
//       <MemoryRouter initialEntries={initialEntries}>
//         {ui}
//         < Route
//           path="*"
//           render={({ history }) => {
//             historyInTest = history;
//             return null;
//           }}
//         />
//       </MemoryRouter>),
//     history: historyInTest
//   };
// };

const renderWithMemoryRouter = (
  ui: React.ReactNode,
  initialRoute: string = "/my/initial-route"
) => {
  const history = createMemoryHistory()
  history.push(initialRoute)
  return {
    ...render(
      <Router history={history}>
        {ui}
      </Router>),
    history
  };
};

export default renderWithMemoryRouter;
