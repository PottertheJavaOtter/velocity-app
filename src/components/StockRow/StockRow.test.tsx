import React from "react";
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import StockRow from "./StockRow";
import renderWithMemoryRouter from "../../mocks/renderers/renderWithMemoryRouter";

describe("StockRow", () => {
  const props = {
    id: "stock-id", ticker: "OKE", name: "ONEOK", shares: 981.2, dividendPerShare: 102.1
  }

  it("should render the stock's information", () => {
    const { container } = render(<StockRow {...props} />);
    expect(container).toMatchSnapshot();
  });

  it("should route to the stock's info page on click", () => {
    const { history } = renderWithMemoryRouter(<StockRow {...props} />);
    const stockRow = screen.getByTestId("stock-row-OKE");
    expect(stockRow.getAttribute("href")).toBe("/stock/stock-id")
    // expect(screen.getByTestId("stock-row-OKE").nodeName).toEqual("A")
    // userEvent.click(screen.getByTestId("stock-row-OKE"))
    // expect(history.location.pathname).toContain("/stock/stock-id");
  });
});
