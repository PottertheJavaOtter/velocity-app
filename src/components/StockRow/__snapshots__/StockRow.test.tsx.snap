// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StockRow should render the stock's information 1`] = `
<div>
  <ion-item
    data-testid="stock-row-OKE"
    href="/stock/stock-id"
  >
    <ion-avatar
      slot="start"
    >
      <img
        src="https://picsum.photos/100/100"
      />
    </ion-avatar>
    <ion-label>
      <h2>
        ONEOK (OKE)
      </h2>
      <p>
        Shares: 
        981.2
      </p>
    </ion-label>
    <ion-note
      color="primary"
      slot="end"
    >
      $100,180.52
    </ion-note>
  </ion-item>
</div>
`;
